File: configuration/src/main/java/com/synopsys/integration/configuration/property/base/NullableProperty.java
package com.synopsys.integration.configuration.property.base;

import org.jetbrains.annotations.NotNull;

import com.synopsys.integration.configuration.parse.ValueParser;

/**
 * A property that returns null when it is not present in a Configuration.
 * @param <V> the type this property returns when it is retrieved from a Configuration.
 */
public abstract class NullableProperty<V, R> extends TypedProperty<V, R> {
    public NullableProperty(@NotNull String key, @NotNull ValueParser<V> parser) {
        super(key, parser);
    }
}

File: configuration/src/main/java/com/synopsys/integration/configuration/property/base/NullableProperty.kt
package com.synopsys.integration.configuration.property.base

import com.synopsys.integration.configuration.parse.ValueParser

abstract class NullableProperty<T>(key: String, parser: ValueParser<T>) : TypedProperty<T>(key, parser) {
    // This is a property with a key and a value, but the value is optional (nullable).

    var replacesDeprecatedProperty: NullableProperty<T>? = null
    fun replacesDeprecatedProperty(oldProperty: NullableProperty<T>): NullableProperty<T> {
        this.replacesDeprecatedProperty = oldProperty
        return this
    }
}